system/lib/wasmfs/wasmfs.cpp

```diff
@@ -6,47 +6,85 @@
 // Current Status: Work in Progress.
 // See https://github.com/emscripten-core/emscripten/issues/15041.
 
+#include "file.h"
+#include "file_table.h"
 #include <emscripten/emscripten.h>
 #include <emscripten/html5.h>
+#include <errno.h>
+#include <mutex>
 #include <stdlib.h>
+#include <utility>
 #include <vector>
 #include <wasi/api.h>
 
 extern "C" {
 
-static std::vector<char> fd_write_stdstream_buffer;
+using namespace wasmfs;
+
+long __syscall_dup2(long oldfd, long newfd) {
+  FileTable::Handle fileTable = FileTable::get();
+
+  auto oldOpenFile = fileTable[oldfd];
+  // If oldfd is not a valid file descriptor, then the call fails,
+  // and newfd is not closed.
+  if (!oldOpenFile) {
+    return -(EBADF);
+  }
+
+  if (newfd < 0) {
+    return -(EBADF);
+  }
+
+  if (oldfd == newfd) {
+    return oldfd;
+  }
+
+  // If the file descriptor newfd was previously open, it will just be overwritten silently.
+  fileTable[newfd] = oldOpenFile;
+  return newfd;
+}
+
+long __syscall_dup(long fd) {
+
+  FileTable::Handle fileTable = FileTable::get();
+
+  // Check that an open file exists corresponding to the given fd.
+  auto currentOpenFile = fileTable[fd];
+  if (!currentOpenFile) {
+    return -(EBADF);
```

Comment:
@sbc100 Thoughts on what to return if there is an invalid fd passed in? According to the manpage, we should return -1. Wondering if I can should be using some macro here? Thanks

https://man7.org/linux/man-pages/man2/dup.2.html#RETURN_VALUE

Reply:
In the `dup2.c` test I am just checking if  `assert(fd == -1)`

Reply:
The man page is c library version of dup which is slightly higher level than this function which is the system call version.

Linux system calls, and therefore musl system calls, and therefore emscripten system calls return `-(errno)` on failure.  The higher level libc function is the in charge of setting errno and returning -1.

Yes it would probably make sense to have a macro here.  Something like `(GET_OPEN_FILE(fd))`;